app/uk/gov/hmrc/incometaxsessiondata/predicates/AuthenticationPredicate.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.incometaxsessiondata.predicates
+
+import play.api.Logging
+import play.api.mvc._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals.{affinityGroup, confidenceLevel}
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.auth.core.{AffinityGroup, AuthorisationException, AuthorisedFunctions, ConfidenceLevel}
+import uk.gov.hmrc.incometaxsessiondata.config.AppConfig
+import uk.gov.hmrc.incometaxsessiondata.connectors.MicroserviceAuthConnector
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+@Singleton
+class AuthenticationPredicate @Inject()(val authConnector: MicroserviceAuthConnector, cc: ControllerComponents,
+                                        val appConfig: AppConfig
+                                       )(implicit ec: ExecutionContext) extends BackendController(cc) with AuthorisedFunctions with Logging {
+
+  val minimumConfidenceLevel: Int = ConfidenceLevel.fromInt(appConfig.confidenceLevel) match {
```

Comment:
consider next approach / I am actually not sure if we can handle exception we raising in this fashion in our BE, so lets raise less exceptions:

```
  private val minimumConfidenceLevel: Option[Int] = ConfidenceLevel.fromInt(appConfig.confidenceLevel).map(_.level).toOption
  
  def async(action: Request[AnyContent] => Future[Result]): Action[AnyContent] =
    Action.async { implicit request =>
      authorised().retrieve(affinityGroup and confidenceLevel) {
        case Some(AffinityGroup.Agent) ~ _ => action(request)
        case _ ~ userConfidence if minimumConfidenceLevel.exists(confidenceLevel => userConfidence.level >= confidenceLevel) =>
          action(request)
        case _ ~ _ =>
          logger.info(s"[AuthenticationPredicate][authenticated] User has confidence level below ${minimumConfidenceLevel}")
          Future(Unauthorized)
      } recover {
```